packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -74,324 +74,465 @@ class FocusNode extends ChangeNotifier {
-  FocusScopeNode _previousSibling;
-
-  FocusScopeNode _firstChild;
-  FocusScopeNode _lastChild;
-
-  FocusNode _focus;
-  List<FocusScopeNode> _focusPath;
-
-  /// Whether this scope is currently active in its parent scope.
-  bool get isFirstFocus => _parent == null || _parent._firstChild == this;
-
-  // Returns this FocusScopeNode's ancestors, starting with the node
-  // below the FocusManager's rootScope.
-  List<FocusScopeNode> _getFocusPath() {
-    final List<FocusScopeNode> nodes = <FocusScopeNode>[this];
-    FocusScopeNode node = _parent;
-    while (node != null && node != _manager?.rootScope) {
-      nodes.add(node);
-      node = node._parent;
-    }
-    return nodes;
-  }
-
-  void _prepend(FocusScopeNode child) {
-    assert(child != this);
-    assert(child != _firstChild);
-    assert(child != _lastChild);
-    assert(child._parent == null);
-    assert(child._manager == null);
-    assert(child._nextSibling == null);
-    assert(child._previousSibling == null);
-    assert(() {
-      FocusScopeNode node = this;
-      while (node._parent != null)
-        node = node._parent;
-      assert(node != child); // indicates we are about to create a cycle
-      return true;
-    }());
-    child._parent = this;
-    child._nextSibling = _firstChild;
-    if (_firstChild != null)
-      _firstChild._previousSibling = child;
-    _firstChild = child;
-    _lastChild ??= child;
-    child._updateManager(_manager);
+    // If this was the focused child for this scope, then unfocus it in that
+    // scope.
+    node.unfocus();
+
+    node._parent = null;
+    _children.remove(node);
+    assert(!_manager.rootScope.descendants.contains(node));
   }
 
+  // Updates the manager for all descendants to the given manager.
   void _updateManager(FocusManager manager) {
-    void update(FocusScopeNode child) {
-      if (child._manager == manager)
-        return;
-      child._manager = manager;
-      // We don't proactively null out the manager for FocusNodes because the
-      // manager holds the currently active focus node until the end of the
-      // microtask, even if that node is detached from the focus tree.
-      if (manager != null)
-        child._focus?._manager = manager;
-      child._visitChildren(update);
+    // TODO(gspencer): There's probably a more efficient way to do this.
+    for (FocusNode node in descendants) {
+      node._manager = manager;
     }
+    _manager = manager;
+  }
 
-    update(this);
+  /// Moves the given node to be a child of this node.
+  ///
+  /// Called whenever the associated widget is rebuilt in order to maintain the
+  /// focus hierarchy.
+  ///
+  /// A widget that manages focus should call this method on the node it owns
+  /// during its `build` method in case the widget is moved from one location
+  /// in the tree to another location that has a different focus scope.
+  @mustCallSuper
+  void reparentIfNeeded(FocusNode child, BuildContext childContext) {
```

Comment:
by convention we always put the context first

Reply:
OK, makes sense.